compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -524,43 +527,80 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
+  /** Refine child based on parent
+   *
+   *  In child class definition, we have:
+   *
+   *      class Child[Ts] extends path.Parent[Us] with Es
+   *      object Child extends path.Parent[Us] with Es
+   *      val child = new path.Parent[Us] with Es           // enum values
+   *
+   *  Given a parent type `parent` and a child symbol `child`, we infer the prefix
+   *  and type parameters for the child:
+   *
+   *      prefix.child[Vs] <:< parent
+   *
+   *  where `Vs` are fresh type variables and `prefix` is the symbol prefix with all
+   *  non-module and non-package `ThisType` replaced by fresh type variables.
+   *
+   *  If the subtyping is true, the instantiated type `p.child[Vs]` is
+   *  returned. Otherwise, `NoType` is returned.
+   *
+   */
+  def refine(parent: Type, child: Symbol): Type = {
+    if (child.isTerm && child.is(Case, butNot = Module)) return child.termRef // enum vals always match
+
+    val childTp  = if (child.isTerm) child.termRef else child.typeRef
+
+    val resTp = instantiate(childTp, expose(parent))(ctx.fresh.setNewTyperState)
+
+    if (!resTp.exists)  {
+      debug.println(s"[refine] unqualified child ousted: ${childTp.show} !< ${parent.show}")
+      NoType
+    }
+    else {
+      debug.println(s"$child instantiated ------> $resTp")
+      resTp
+    }
+  }
+
+  /** Instantiate type `tp1` to be a subtype of `tp2`
+   *
+   *  Return the instantiated type if type parameters and this type
+   *  in `tp1` can be instantiated such that `tp1 <:< tp2`.
    *
-   *  E.g. if `tp1` is `Option[Int]`, `tp2` is `Some`, then return
-   *  `Some[Int]`.
+   *  Otherwise, return NoType.
    *
-   *  If `tp1` is `path1.A`, `tp2` is `path2.B`, and `path1` is subtype of
-   *  `path2`, then return `path1.B`.
    */
-  def refine(tp1: Type, tp2: Type): Type = (tp1, tp2) match {
-    case (tp1: RefinedType, _: TypeRef) => tp1.wrapIfMember(refine(tp1.parent, tp2))
-    case (tp1: HKApply, _) => refine(tp1.superType, tp2)
-    case (TypeRef(ref1: TypeProxy, _), tp2 @ TypeRef(ref2: TypeProxy, _)) =>
-      if (ref1.underlying <:< ref2.underlying) tp2.derivedSelect(ref1) else tp2
-    case (TypeRef(ref1: TypeProxy, _), tp2 @ TermRef(ref2: TypeProxy, _)) =>
-      if (ref1.underlying <:< ref2.underlying) tp2.derivedSelect(ref1) else tp2
-    case _ => tp2
+  def instantiate(tp1: Type, tp2: Type)(implicit ctx: Context): Type = {
+    val approx = new TypeMap {
+      def apply(t: Type): Type = t match {
+        case t @ ThisType(tref) if !tref.symbol.isStaticOwner && !tref.symbol.is(Module)  =>
+          newTypeVar(TypeBounds.upper(mapOver(tref & tref.givenSelfType)))
+        case _ =>
+          mapOver(t)
+      }
+    }
+
+    val tvars = tp1.typeParams.map { tparam => newTypeVar(tparam.paramInfo.bounds) }
+    val protoTp1 = approx(tp1.appliedTo(tvars))
+
+    if (protoTp1 <:< tp2 && isFullyDefined(protoTp1, ForceDegree.all)) protoTp1
+    else {
+      debug.println(s"$protoTp1 <:< $tp2 = false")
+      NoType
```

Comment:
Neat that this works! We might want to move it to something more generally available since it looks like the Linker will need something like this was well /cc @nicolasstucki. @DarkDimius